storagemarket/lp2pimpl/net.go

```diff
@@ -386,3 +396,173 @@ func (p *DealProvider) getDealStatus(req types.DealStatusRequest, reqLog *zap.Su
 		NBytesReceived: bts,
 	}
 }
+
+func (p *DealProvider) handleLegacyDealStream(s network.Stream) {
+	start := time.Now()
+	reqLogUuid := uuid.New()
+	reqLog := log.With("reqlog-uuid", reqLogUuid.String(), "client-peer", s.Conn().RemotePeer())
+	reqLog.Debugw("new legacy deal status request")
+	defer func() {
+		err := s.Close()
+		if err != nil {
+			reqLog.Infow("closing stream", "err", err)
+		}
+		reqLog.Debugw("handled legacy deal status request", "duration", time.Since(start).String())
+	}()
+
+	_ = s.SetReadDeadline(time.Now().Add(providerReadDeadline))
+
+	ts, err := p.fullNode.ChainHead(p.ctx)
+	if err != nil {
+		reqLog.Errorf("could not get chain head: %s", err)
+	}
+
+	maddr, err := p.spApi.ActorAddress(p.ctx)
+	if err != nil {
+		reqLog.Errorf("could not get miner actor address: %s", err)
+	}
+
+	mi, err := p.fullNode.StateMinerInfo(p.ctx, maddr, ts.Key())
+	if err != nil {
+		reqLog.Errorf("could not get miner info: %s", err)
+	}
+
+	switch s.Protocol() {
+	case gfm_storagemarket.DealProtocolID101:
```

Comment:
I tried to save some repetitive code but that just make is more difficult to understand the execution. Since all 3 types have different Proposal type, it becomes a little tricky. The code which is common is in the middle of the uncommon code ends. I think it would be better to leave it like this. More neat. 
Let me know if that is not the way to go and I will try to find a way to shorten it.

Reply:
Yes it's not so straight-forward - I refactored a little bit, I think it's good enough for now.